File: wire-schema/src/test/java/com/squareup/wire/schema/OptionsTest.java
/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.schema;

import com.google.common.collect.ImmutableMap;
import java.util.LinkedHashSet;
import java.util.Set;
import org.junit.Test;

import static java.util.Arrays.asList;
import static org.assertj.core.api.Assertions.assertThat;

public final class OptionsTest {
  @Test public void structuredAndUnstructuredOptions() throws Exception {
    // From https://developers.google.com/protocol-buffers/docs/proto#options
    Schema schema = new SchemaBuilder()
        .add("foo.proto", ""
            + "import \"google/protobuf/descriptor.proto\";\n"
            + "message FooOptions {\n"
            + "  optional int32 opt1 = 1;\n"
            + "  optional string opt2 = 2;\n"
            + "}\n"
            + "\n"
            + "extend google.protobuf.FieldOptions {\n"
            + "  optional FooOptions foo_options = 1234;\n"
            + "}\n"
            + "\n"
            + "message Bar {\n"
            + "  optional int32 a = 1 [(foo_options).opt1 = 123, (foo_options).opt2 = \"baz\"];\n"
            + "  optional int32 b = 2 [(foo_options) = { opt1: 456 opt2: \"quux\" }];\n"
            + "}\n")
        .build();

    ProtoMember fooOptions = ProtoMember.get(Options.FIELD_OPTIONS, "foo_options");
    ProtoMember opt1 = ProtoMember.get(ProtoType.get("FooOptions"), "opt1");
    ProtoMember opt2 = ProtoMember.get(ProtoType.get("FooOptions"), "opt2");

    MessageType bar = (MessageType) schema.getType("Bar");
    assertThat(bar.field("a").options().map()).isEqualTo(ImmutableMap.of(
        fooOptions, ImmutableMap.of(opt1, "123", opt2, "baz")));
    assertThat(bar.field("b").options().map()).isEqualTo(ImmutableMap.of(
        fooOptions, ImmutableMap.of(opt1, "456", opt2, "quux")));
  }

  @Test public void textFormatCanOmitMapValueSeparator() throws Exception {
    Schema schema = new SchemaBuilder()
        .add("foo.proto", ""
            + "import \"google/protobuf/descriptor.proto\";\n"
            + "message FooOptions {\n"
            + "  optional BarOptions bar = 2;\n"
            + "}\n"
            + "message BarOptions {\n"
            + "  optional int32 baz = 2;\n"
            + "}\n"
            + "\n"
            + "extend google.protobuf.FieldOptions {\n"
            + "  optional FooOptions foo = 1234;\n"
            + "}\n"
            + "\n"
            + "message Message {\n"
            + "  optional int32 b = 2 [(foo) = { bar { baz: 123 } }];\n"
            + "}\n")
        .build();

    ProtoMember foo = ProtoMember.get(Options.FIELD_OPTIONS, "foo");
    ProtoMember bar = ProtoMember.get(ProtoType.get("FooOptions"), "bar");
    ProtoMember baz = ProtoMember.get(ProtoType.get("BarOptions"), "baz");

    MessageType message = (MessageType) schema.getType("Message");
    assertThat(message.field("b").options().map()).isEqualTo(ImmutableMap.of(
        foo, ImmutableMap.of(bar, ImmutableMap.of(baz, "123"))));
  }

  @Test public void fullyQualifiedOptionFields() throws Exception {
    Schema schema = new SchemaBuilder()
        .add("a/b/more_options.proto", ""
            + "syntax = \"proto2\";\n"
            + "package a.b;\n"
            + "\n"
            + "import \"google/protobuf/descriptor.proto\";\n"
            + "\n"
            + "extend google.protobuf.MessageOptions {\n"
            + "  optional MoreOptions more_options = 17000;\n"
            + "}\n"
            + "\n"
            + "message MoreOptions {\n"
            + "  extensions 100 to 200;\n"
            + "}\n")
        .add("a/c/event_more_options.proto", ""
            + "syntax = \"proto2\";\n"
            + "package a.c;\n"
            + "\n"
            + "import \"a/b/more_options.proto\";\n"
            + "\n"
            + "extend a.b.MoreOptions {\n"
            + "  optional EvenMoreOptions even_more_options = 100;\n"
            + "}\n"
            + "\n"
            + "message EvenMoreOptions {\n"
            + "  optional string string_option = 1;\n"
            + "}\n")
        .add("a/d/message.proto", ""
            + "syntax = \"proto2\";\n"
            + "package a.d;\n"
            + "\n"
            + "import \"a/b/more_options.proto\";\n"
            + "import \"a/c/event_more_options.proto\";\n"
            + "\n"
            + "message Message {\n"
            + "  option (a.b.more_options) = {\n"
            + "    [a.c.even_more_options]: {string_option: \"foo\"}\n"
            + "  };\n"
            + "}\n")
        .build();
    ProtoType moreOptionsType = ProtoType.get("a.b.MoreOptions");
    ProtoType evenMoreOptionsType = ProtoType.get("a.c.EvenMoreOptions");
    ProtoMember moreOptions = ProtoMember.get(Options.MESSAGE_OPTIONS, "a.b.more_options");
    ProtoMember evenMoreOptions = ProtoMember.get(moreOptionsType, "a.c.even_more_options");
    ProtoMember stringOption = ProtoMember.get(evenMoreOptionsType, "string_option");
    MessageType message = (MessageType) schema.getType("a.d.Message");
    assertThat(message.options().map()).isEqualTo(
        ImmutableMap.of(moreOptions, ImmutableMap.of(
            evenMoreOptions, ImmutableMap.of(stringOption, "foo"))));
  }

  @Test public void resolveFieldPathMatchesFirstSegment() throws Exception {
    assertThat(asList(Options.resolveFieldPath("a.b.c.d", set("a", "z", "y"))))
        .containsExactly("a", "b", "c", "d");
  }

  @Test public void resolveFieldPathMatchesMultipleSegments() throws Exception {
    assertThat(asList(Options.resolveFieldPath("a.b.c.d", set("a.b", "z.b", "y.b"))))
        .containsExactly("a.b", "c", "d");
  }

  @Test public void resolveFieldPathMatchesAllSegments() throws Exception {
    assertThat(asList(Options.resolveFieldPath("a.b.c.d", set("a.b.c.d", "z.b.c.d"))))
        .containsExactly("a.b.c.d");
  }

  @Test public void resolveFieldPathMatchesOnlySegment() throws Exception {
    assertThat(asList(Options.resolveFieldPath("a", set("a", "b")))).containsExactly("a");
  }

  @Test public void resolveFieldPathDoesntMatch() throws Exception {
    assertThat(Options.resolveFieldPath("a.b", set("c", "d"))).isNull();
  }

  private Set<String> set(String... elements) {
    return new LinkedHashSet<>(asList(elements));
  }
}

File: wire-schema/src/test/java/com/squareup/wire/schema/OptionsTest.kt
/*
 * Copyright (C) 2015 Square, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package com.squareup.wire.schema

import org.assertj.core.api.Assertions.assertThat
import org.junit.Test

class OptionsTest {
  @Test
  fun structuredAndUnstructuredOptions() {
    // From https://developers.google.com/protocol-buffers/docs/proto#options
    val schema = RepoBuilder()
        .add("foo.proto",
            """
            |import "google/protobuf/descriptor.proto";
            |message FooOptions {
            |  optional int32 opt1 = 1;
            |  optional string opt2 = 2;
            |}
            |
            |extend google.protobuf.FieldOptions {
            |  optional FooOptions foo_options = 1234;
            |}
            |
            |message Bar {
            |  optional int32 a = 1 [(foo_options).opt1 = 123, (foo_options).opt2 = "baz"];
            |  optional int32 b = 2 [(foo_options) = { opt1: 456 opt2: "quux" }];
            |}
            """.trimMargin()
        )
        .schema()

    val fooOptions = ProtoMember.get(Options.FIELD_OPTIONS, "foo_options")
    val opt1 = ProtoMember.get(ProtoType.get("FooOptions"), "opt1")
    val opt2 = ProtoMember.get(ProtoType.get("FooOptions"), "opt2")

    val bar = schema.getType("Bar") as MessageType
    assertThat(bar.field("a")!!.options.map)
        .isEqualTo(mapOf(fooOptions to mapOf(opt1 to "123", opt2 to "baz")))
    assertThat(bar.field("b")!!.options.map)
        .isEqualTo(mapOf(fooOptions to mapOf(opt1 to "456", opt2 to "quux")))
  }

  @Test
  fun textFormatCanOmitMapValueSeparator() {
    val schema = RepoBuilder()
        .add("foo.proto",
            """
            |import "google/protobuf/descriptor.proto";
            |message FooOptions {
            |  optional BarOptions bar = 2;
            |}
            |message BarOptions {
            |  optional int32 baz = 2;
            |}
            |
            |extend google.protobuf.FieldOptions {
            |  optional FooOptions foo = 1234;
            |}
            |
            |message Message {
            |  optional int32 b = 2 [(foo) = { bar { baz: 123 } }];
            |}
            """.trimMargin()
        )
        .schema()

    val foo = ProtoMember.get(Options.FIELD_OPTIONS, "foo")
    val bar = ProtoMember.get(ProtoType.get("FooOptions"), "bar")
    val baz = ProtoMember.get(ProtoType.get("BarOptions"), "baz")

    val message = schema.getType("Message") as MessageType
    assertThat(message.field("b")!!.options.map)
        .isEqualTo(mapOf(foo to mapOf(bar to mapOf(baz to "123"))))
  }

  @Test
  fun fullyQualifiedOptionFields() {
    val schema = RepoBuilder()
        .add("a/b/more_options.proto",
            """
            |syntax = "proto2";
            |package a.b;
            |
            |import "google/protobuf/descriptor.proto";
            |
            |extend google.protobuf.MessageOptions {
            |  optional MoreOptions more_options = 17000;
            |}
            |
            |message MoreOptions {
            |  extensions 100 to 200;
            |}
            """.trimMargin()
        )
        .add("a/c/event_more_options.proto",
            """
            |syntax = "proto2";
            |package a.c;
            |
            |import "a/b/more_options.proto";
            |
            |extend a.b.MoreOptions {
            |  optional EvenMoreOptions even_more_options = 100;
            |}
            |
            |message EvenMoreOptions {
            |  optional string string_option = 1;
            |}
            """.trimMargin())
        .add("a/d/message.proto",
            """
            |syntax = "proto2";
            |package a.d;
            |
            |import "a/b/more_options.proto";
            |import "a/c/event_more_options.proto";
            |
            |message Message {
            |  option (a.b.more_options) = {
            |    [a.c.even_more_options]: {string_option: "foo"}
            |  };
            |}
            """.trimMargin()
        )
        .schema()
    val moreOptionsType = ProtoType.get("a.b.MoreOptions")
    val evenMoreOptionsType = ProtoType.get("a.c.EvenMoreOptions")
    val moreOptions = ProtoMember.get(Options.MESSAGE_OPTIONS, "a.b.more_options")
    val evenMoreOptions = ProtoMember.get(moreOptionsType, "a.c.even_more_options")
    val stringOption = ProtoMember.get(evenMoreOptionsType, "string_option")
    val message = schema.getType("a.d.Message") as MessageType

    assertThat(message.options.map)
        .isEqualTo(mapOf(moreOptions to mapOf(evenMoreOptions to mapOf(stringOption to "foo"))))
  }

  @Test
  fun resolveFieldPathMatchesFirstSegment() {
    assertThat(Options.resolveFieldPath("a.b.c.d", setOf("a", "z", "y")))
        .containsExactly("a", "b", "c", "d")
  }

  @Test
  fun resolveFieldPathMatchesMultipleSegments() {
    assertThat(Options.resolveFieldPath("a.b.c.d", setOf("a.b", "z.b", "y.b")))
        .containsExactly("a.b", "c", "d")
  }

  @Test
  fun resolveFieldPathMatchesAllSegments() {
    assertThat(Options.resolveFieldPath("a.b.c.d", setOf("a.b.c.d", "z.b.c.d")))
        .containsExactly("a.b.c.d")
  }

  @Test
  fun resolveFieldPathMatchesOnlySegment() {
    assertThat(Options.resolveFieldPath("a", setOf("a", "b"))).containsExactly("a")
  }

  @Test
  fun resolveFieldPathDoesntMatch() {
    assertThat(Options.resolveFieldPath("a.b", setOf("c", "d"))).isNull()
  }
}
